apps/rush-lib/src/cli/test/RushCommandLineParser.test.ts

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+// Mock child_process so we can verify tasks are (or are not) invoked as we expect
+jest.mock('child_process');
+// Mock RushCommandLineParser itself to prevent `process.exit` to be called on failure
+jest.mock('../RushCommandLineParser', rushCommandLineParserMockFactory);
+
+import { resolve } from 'path';
+import { chdir } from 'process';
+import { ChildProcessModuleMock, ISpawnMockConfig } from 'child_process';
+import { FileSystem } from '@microsoft/node-core-library';
+import { Interleaver } from '@microsoft/stream-collator';
+import { RushCommandLineParser } from '../RushCommandLineParser';
+
+/**
+ * Interface definition for a test instance for the RushCommandLineParser.
+ */
+interface IParserTestInstance {
+  parser: RushCommandLineParser;
+  spawnMock: jest.Mock;
+}
+
+/**
+ * Helper to set up a test instance for RushCommandLineParser.
+ */
+function newCommandLineParserInstance(repoName: string, taskName: string): IParserTestInstance {
+  // Point to the test repo folder
+  chdir(resolve(__dirname, repoName));
```

Comment:
I suspect that `chdir` affects the global execution environment for all tests (and maybe even other tests running concurrently?).  Is there a way to avoid doing relying on `chdir`? If not, perhaps the test should restore the old cwd as part of its cleanup. @iclanton

Reply:
Empirically this hasn't caused a problem (so far). But I agree it is brittle. And I can certainly restore the old cwd as part of cleanup.

An alternative is to follow inversion-of-control pattern and support passing in an optional `startPath` argument to the constructor of `RushCommandLineParser` which the unit tests can use to indicate where the repo is. The absence of that optional argument would follow the current logic of getting `process.cwd`. (I prefer this solution, btw.)

To make that change, I would need to change the signature of `RushConfiguration.tryFindRushJsonLocation`. What are the breaking-changing rules I need to pay attention to?

Like from this:

```JavaScript
  public static tryFindRushJsonLocation(verbose: boolean = true): string | undefined {
    let currentFolder: string = process.cwd();
```

to this:

```JavaScript
  public static tryFindRushJsonLocation(startPath?: string, verbose: boolean = true): string | undefined {
    let currentFolder: string = startPath === undefined ? process.cwd() : startPath;
```

or this:

```JavaScript
  public static tryFindRushJsonLocation(verbose: boolean = true, startPath?: string): string | undefined {
    let currentFolder: string = startPath === undefined ? process.cwd() : startPath;
```

Questions:
1. From a breaking-change perspective, is there a preference on whether the additional parameter is prepended vs. appended to the argument list? (Here I am proposing it comes before the `verbose` arg because it typically seems appropriate that path-ish arguments precede verbosity arguments.) Or I can append the `startPath` as the 2nd arg to minimize the breaking change. Or change to a config arg pattern. Or make it more complex like so:

```JavaScript
  public static tryFindRushJsonLocation(startPath?: string, verbose?: boolean): string | undefined;
  public static tryFindRushJsonLocation(verbose?: boolean): string | undefined;
  public static tryFindRushJsonLocation(verboseOrStartPath?: boolean | string, verboseOrNot?: boolean): string | undefined {
    const startPath: string = typeof verboseOrStartPath === 'string' ? verboseOrStartPath : process.cwd();
    const verbose: boolean = typeof verboseOrStartPath === 'boolean'
      ? verboseOrStartPath
      : verboseOrNot !== undefined
        ? verboseOrNot
        : true;
    let currentFolder: string = startPath;
```

But that's a lot of complex code which might not be necessary if compatibility of this API is not paramount.

2. What is the protocol for making changes to this API wrt `~/common/reviews/api/rush-lib.api.ts`?

3. Or should I introduce a new static API that supports both args and leave `tryFindRushJsonLocation` untouched? If you prefer this approach, do you have naming convention for subsequent function names?

Reply:
1. That file is generated and so will be overwritten once you do the build after a breaking change. You should check in whatever changes are made to that file.

2. This particular API is not heavily used outside the Rush codebase . Most of the time people are attempting to load the Rush config file and use `loadFromDefaultLocation()`, so I would be okay with adding an optional string path argument to the front. It technically will be a breaking change.

---
In reply to: [253222932](https://github.com/Microsoft/web-build-tools/pull/1062#discussion_r253222932) [](ancestors = 253222932)

Reply:
If we're going to break it, my preference would be to convert the contract to an options interface, something like
```ts
  /** 
   * Options for RushConfiguration.tryFindRushJsonLocation().
   * @public 
   */
  export interface ITryFindRushJsonOptions {
    showWarnings?: boolean; // defaults to false, inverse of the old "verbose"
    startingFolder?: string; // defaults to cwd
  }
  public static tryFindRushJsonLocation(options?: ITryFindRushJsonOptions): string | undefined {
  }
```

The `verbose` option is poorly conceived and its default is backwards, so long term I'd like to remove it. But for now we can rename it to `showWarnings`.

I agree with @nickpape-msft that probably Rush is the only code base that's passing any parameters to this API today.  If you wanted to be paranoid, you could do something like:

```ts
if (typeof options === 'boolean') {
  options = { showWarnings: !(options as boolean) };
}
```